File: csharp/bob/Bob.cs
using System;

namespace bob
{
	public class Bob
	{
		public static string Hey (string message)
		{
			if (IsEmpty (message)) {
				return "Fine. Be that way!";
			}
			if (IsYelling (message)) {
				return "Whoa, chill out!";
			}
			if (IsQuestion (message)) {
				return "Sure.";
			}
			return "Whatever.";
		}

		private static bool IsEmpty (string message)
		{
			return string.IsNullOrWhiteSpace(message.Trim());
		}

		private static bool IsYelling (string message)
		{
			return message.ToUpper () == message && message.ToLower () != message;
		}

		private static bool IsQuestion (string message)
		{
			return message.Trim().EndsWith("?");
		}
	}
}

File: ruby/bob/bob.rb
class Bob
  def hey(remark)
    return "Fine. Be that way!" if silence?(remark)
    return "Whoa, chill out!" if shout?(remark)
    return "Sure." if question?(remark)
    "Whatever."
  end

  private

  def question?(remark)
    remark.end_with?("?")
  end

  def shout?(remark)
    remark == remark.upcase && remark =~ /[A-Z]/
  end

  def silence?(remark)
    remark.strip.empty?
  end
end